src/main/java/emu/grasscutter/game/managers/EnergyManager.java

```diff
@@ -0,0 +1,215 @@
+import emu.grasscutter.net.proto.PropChangeReasonOuterClass.PropChangeReason;
+import emu.grasscutter.server.game.GameSession;
+import emu.grasscutter.utils.Position;
+
+import static emu.grasscutter.Configuration.GAME_OPTIONS;
+
+import java.util.Optional;
+
+import com.google.protobuf.InvalidProtocolBufferException;
+
+public class EnergyManager {
+    private final Player player;
+    
+    public EnergyManager(Player player) {
+        this.player = player;
+    }
+
+    public Player getPlayer() {
+        return this.player;
+    }
+
+    /**********
+        Particle creation for elemental skills.
+    **********/
+    private int getCastingAvatarIdForElemBall(int invokeEntityId) {
+		// To determine the avatar that has cast the skill that caused the energy particle to be generated,
+		// we have to look at the entity that has invoked the ability. This can either be that avatar directly, 
+		// or it can be an `EntityClientGadget`, owned (some way up the owner hierarchy) by the avatar 
+		// that cast the skill.
+		int res = 0;
+
+		// Try to get the invoking entity from the scene.
+		GameEntity entity = player.getScene().getEntityById(invokeEntityId);
+
+		// If this entity is null, or not an `EntityClientGadget`, we assume that we are directly 
+		// looking at the casting avatar (the null case will happen if the avatar was switched out 
+		// between casting the skill and the particle being generated).
+		if (!(entity instanceof EntityClientGadget)) {
+			res = invokeEntityId;
+		}
+		// If the entity is a `EntityClientGadget`, we need to "walk up" the owner hierarchy,
+		// until the owner is no longer a gadget. This should then be the ID of the casting avatar.
+		else {	
+			while (entity instanceof EntityClientGadget gadget) {
+				res = gadget.getOwnerEntityId();
+				entity = player.getScene().getEntityById(gadget.getOwnerEntityId());
+			}
+		}
+		
+		return res;
+	}
+    
+    public void handleGenerateElemBall(AbilityInvokeEntry invoke) throws InvalidProtocolBufferException {
+		// ToDo: 
+		// This is also called when a weapon like Favonius Warbow etc. creates energy through its passive.
+		// We are not handling this correctly at the moment.
+
+		// Get action info.
+		AbilityActionGenerateElemBall action = AbilityActionGenerateElemBall.parseFrom(invoke.getAbilityData());
+		if (action == null) {
+			return;
+		}
+		
+		// Determine the element of the energy particle that we have to generate.
+		// In case we can't, we default to an elementless particle.
+		// The element is the element of the avatar that has cast the ability.
+		// We can get that from the avatar's skill depot.
+		int itemId = 2024;
+		
+		// Try to fetch the avatar from the player's party and determine their element.
+		// ToDo: Does this work in co-op?
+		int avatarId = getCastingAvatarIdForElemBall(invoke.getEntityId());	
+		Optional<EntityAvatar> avatarEntity = player.getTeamManager().getActiveTeam()
```

Comment:
It is reccomended to use `TeamManager::getCurrentAvatarEntity()`

Reply:
But that only gives me the currently active avatar, doesn't it? The character generating the particle doesn't have to be active (think for example Zhongli's pillar generating a particle while he is off field).